server/resourcemanager/resource_manager.go

```diff
@@ -0,0 +1,347 @@
+	Type string `mapstructure:"type"`
+	Spec T      `mapstructure:"spec"`
+}
+
+type ResourceHandler[T ResourceSpec] interface {
+	SetID(T, id.ID) T
+	List(_ context.Context, take, skip int, query, sortBy, sortDirection string) ([]T, error)
+	Count(_ context.Context, query string) (int, error)
+	Create(context.Context, T) (T, error)
+	Update(context.Context, T) (T, error)
+	Get(context.Context, id.ID) (T, error)
+	Delete(context.Context, id.ID) error
+}
+
+type manager[T ResourceSpec] struct {
+	resourceType string
+	handler      ResourceHandler[T]
+	idgen        func() id.ID
+}
+
+func New[T ResourceSpec](resourceType string, handler ResourceHandler[T], idgenFn func() id.ID) *manager[T] {
+	return &manager[T]{
+		resourceType: resourceType,
+		handler:      handler,
+		idgen:        idgenFn,
+	}
+}
+
+func (m *manager[T]) RegisterRoutes(r *mux.Router) *mux.Router {
+	// prefix is /{resourceType | lowercase}/
+	subrouter := r.PathPrefix("/" + strings.ToLower(m.resourceType)).Subrouter()
+
+	subrouter.HandleFunc("/", m.list).Methods(http.MethodGet).Name(fmt.Sprintf("%s.List", m.resourceType))
+	subrouter.HandleFunc("/", m.create).Methods(http.MethodPost).Name(fmt.Sprintf("%s.Create", m.resourceType))
+	subrouter.HandleFunc("/", m.update).Methods(http.MethodPut).Name(fmt.Sprintf("%s.Update", m.resourceType))
+
+	subrouter.HandleFunc("/{id}", m.get).Methods(http.MethodGet).Name(fmt.Sprintf("%s.Get", m.resourceType))
+	subrouter.HandleFunc("/{id}", m.delete).Methods(http.MethodDelete).Name(fmt.Sprintf("%s.Delete", m.resourceType))
+
+	return subrouter
+}
+
+func (m *manager[T]) create(w http.ResponseWriter, r *http.Request) {
+	m.operationWithBody(w, r, http.StatusCreated, "creating", m.handler.Create)
+}
+
+func (m *manager[T]) update(w http.ResponseWriter, r *http.Request) {
+	m.operationWithBody(w, r, http.StatusOK, "updating", m.handler.Update)
+}
+
+func getIntFromQuery(r *http.Request, key string) (int, error) {
+	str := r.URL.Query().Get(key)
+	if str == "" {
+		return 0, nil
+	}
+
+	val, err := strconv.Atoi(str)
+	if err != nil {
+		return 0, fmt.Errorf("'%s' is not a number", str)
+	}
+
+	return val, nil
+}
+
+func paginationParams(r *http.Request) (take, skip int, query, sortBy, sortDirection string, err error) {
+	take, err = getIntFromQuery(r, "take")
+	if err != nil {
+		err = fmt.Errorf("error reading take param: %w", err)
+		return
+	}
+
+	skip, err = getIntFromQuery(r, "skip")
+	if err != nil {
+		err = fmt.Errorf("error reading skip param: %w", err)
+		return
+	}
+
+	sortBy = r.URL.Query().Get("sortBy")
+	sortDirection = r.URL.Query().Get("sortDirection")
+
+	query = r.URL.Query().Get("query")
+
+	return
+}
+
+func (m *manager[T]) list(w http.ResponseWriter, r *http.Request) {
+	encoder, err := encoderFromRequest(r)
+	if err != nil {
+		writeResponse(w, http.StatusBadRequest, fmt.Sprintf("cannot process request: %s", err.Error()))
+		return
+	}
+	w.Header().Set("Content-Type", encoder.ResponseContentType())
+
+	ctx := r.Context()
```

Comment:
It's better to use the context from the request, so if there's any instrumentation, it gets propagated.

Reply:
`r` is the request actually

Reply:
I have no idea why, but I read it as `ctx := ctx.Background()`. I must be tired :grimacing: